inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/schema/IcebergSchemaChangeHelper.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.inlong.sort.protocol.enums.SchemaChangePolicy;
+import org.apache.inlong.sort.protocol.enums.SchemaChangeType;
+import org.apache.inlong.sort.schema.TableChange;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.JsonNode;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.Transaction;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.flink.FlinkSchemaUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Iceberg schema change helper
+ * */
+public class IcebergSchemaChangeHelper extends SchemaChangeHelper {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(IcebergSchemaChangeHelper.class);
+    private transient Catalog catalog;
+
+    private transient SupportsNamespaces asNamespaceCatalog;
+
+    public IcebergSchemaChangeHelper(JsonDynamicSchemaFormat dynamicSchemaFormat, boolean schemaChange,
+            Map<SchemaChangeType, SchemaChangePolicy> policyMap, String databasePattern, String tablePattern,
+            SchemaUpdateExceptionPolicy exceptionPolicy,
+            SinkTableMetricData metricData, DirtySinkHelper<Object> dirtySinkHelper,
+            Catalog catalog,
+            SupportsNamespaces asNamespaceCatalog) {
+        super(dynamicSchemaFormat, schemaChange, policyMap, databasePattern,
+                tablePattern, exceptionPolicy, metricData, dirtySinkHelper);
+        this.catalog = catalog;
+        this.asNamespaceCatalog = asNamespaceCatalog;
+    }
+    @Override
+    public void doAlterOperation(String database, String table, byte[] originData, String originSchema, JsonNode data,
+            Map<SchemaChangeType, List<AlterColumn>> typeMap) {
+        for (Map.Entry<SchemaChangeType, List<AlterColumn>> kv : typeMap.entrySet()) {
+            SchemaChangePolicy policy = policyMap.get(kv.getKey());
+            doSchemaChangeBase(kv.getKey(), policy, originSchema);
+            if (policy == SchemaChangePolicy.ENABLE) {
+                try {
+                    switch (kv.getKey()) {
+                        case ADD_COLUMN:
+                            doAddColumn(kv.getValue(), TableIdentifier.of(database, table));
+                            break;
+                        case DROP_COLUMN:
+                            doDropColumn(kv.getKey(), originSchema);
+                            break;
+                        case RENAME_COLUMN:
+                            doRenameColumn(kv.getKey(), originSchema);
+                            break;
+                        case CHANGE_COLUMN_TYPE:
+                            doChangeColumnType(kv.getKey(), originSchema);
```

Comment:
doSchemaChangeBase is called before and also inside doChangeColumnType, why is that and any way to improve this ?

Reply:
Fixed